### PR TITLE
Add Storage Dashboard visualizations for histories

### PR DESCRIPTION
**Pull request by davelopez**
First simple visualizations (Bar Charts) of the disk space taken by your histories. The charts are integrated into the Storage Dashboard to help to have an overview of the space taken and potentially recoverable.

You can have a general overview of all your histories or an overview of one particular history and its datasets. Some charts are interactive and you can select the different entries to get more information or perform some actions.

When clicking on the `History Size` icon in the History counter, now you will get directed to that particular history storage overview. You can still access the main Storage Dashboard page from the Quota meter at the top right of the page.

https://user-images.githubusercontent.com/46503462/233411615-7cfe8cc2-7f98-42be-841f-a1bb4704b1d3.mp4

## TODO
- [x] Adapt the StorageDashboardRouter to work like other sub-routers. Similar to https://github.com/galaxyproject/galaxy/pull/14445
- [x] Add charts visualizing storage size for all histories (HistoriesStorageOverview.vue)
- [x] Add charts visualizing storage size for the contents of a particular history (HistoryStorageOverview.vue)
- [ ] Get history name from the store (requires #15108)

## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] Instructions for manual testing are as follows:
  - Go to `http://127.0.0.1:8080/storage/histories` and visually explore your data

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

**Comment by davelopez** [comment]
Thank you @afgane for your feedback!

Those are great ideas, I will definitely implement them :+1:

**Comment by afgane** [comment]
The bar graph looks really nice. Well done.

(now if we could also have a Runtime Dashboard that shows runtime of jobs within a history, that'd be neat ;) )

**Comment by davelopez** [comment]
> (now if we could also have a Runtime Dashboard that shows runtime of jobs within a history, that'd be neat ;) )

That sounds like a nice idea! but definitely for a different PR :laughing:

File: client/src/components/User/DiskUsage/Visualizations/Charts/utils.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import { bytesToString } from "@/utils/utils";
+import type { DataValuePoint } from ".";
+
+export function bytesLabelFormatter(dataPoint?: DataValuePoint | null): string {
+    return dataPoint ? `${dataPoint.label}: ${bytesToString(dataPoint.value)}` : "No data";
+}
+
+export function bytesValueFormatter(value: number): string {
+    return bytesToString(value);
+}
```
Comment (ElectronicBlueberry) [inline client/src/components/User/DiskUsage/Visualizations/Charts/utils.ts]: `formatters` or `formating` may be more fitting module names than `utils`.
I'd like to keep the amount of `util` modules in the project to a minimum, as a module with such a name could contain anything which tends to make code harder to logically follow.

File: client/src/components/User/DiskUsage/Visualizations/HistoriesStorageOverview.vue [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+        }
+    } catch (error) {
+        errorToast(`${error}`, localize("An error occurred while permanently deleting the history."));
+    }
+}
+</script>
+<template>
+    <div class="mx-3">
+        <router-link :to="{ name: 'StorageDashboard' }">{{ localize("Back to Dashboard") }}</router-link>
+        <h2 class="text-center my-3">
```
Comment (ElectronicBlueberry) [inline client/src/components/User/DiskUsage/Visualizations/HistoriesStorageOverview.vue]: This is a top level heading and should therefore be h1. Consider using the `Heading` component here

File: client/src/components/User/DiskUsage/Visualizations/HistoryStorageOverview.vue [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+        }
+    } catch (error) {
+        errorToast(`${error}`, localize("An error occurred while permanently deleting the dataset."));
+    }
+}
+</script>
+<template>
+    <div class="mx-3">
+        <router-link :to="{ name: 'StorageDashboard' }">{{ localize("Back to Dashboard") }}</router-link>
+        <h2 class="text-center my-3">
```
Comment (ElectronicBlueberry) [inline client/src/components/User/DiskUsage/Visualizations/HistoryStorageOverview.vue]: This is a top level heading and should therefore be h1. Consider using the `Heading` component here

File: client/src/components/User/DiskUsage/Visualizations/service.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
+import { fetcher } from "@/schema/fetcher";
+
+const _getHistories = fetcher.path("/api/histories").method("get").create();
+const _getDatasets = fetcher.path("/api/datasets").method("get").create();
+const _undeleteHistory = fetcher.path("/api/histories/deleted/{history_id}/undelete").method("post").create();
+const _purgeHistory = fetcher.path("/api/histories/{history_id}").method("delete").create();
+const _updateDataset = fetcher.path("/api/histories/{history_id}/contents/{type}s/{id}").method("put").create();
+const _purgeDataset = fetcher.path("/api/histories/{history_id}/contents/{type}s/{id}").method("delete").create();
```
Comment (ElectronicBlueberry) [inline client/src/components/User/DiskUsage/Visualizations/service.ts]: Prefixing something with `_` per js convention usually indicates that it is unused/discarded. Please remove these prefixes.

**Review by ElectronicBlueberry — COMMENTED** [review]
This looks very informative, well done!

General feedback:

The bar charts y axis labels are a bit small.

Would it be possible to indicate if a history is deleted/recoverable storage space within the main bar chart view, without hovering, to make them more obvious?

image
It can be quite challenging to hover over these smaller bars. would it be possible to trigger the tooltip on hovering over the column?

image
I find this link a bit too hidden. Maybe add a small arrow to the left, or change it to a button

**Review by dannon — APPROVED** [review]

**Comment by davelopez** [comment]
Thank you for the reviews! I'll go and make the requested changes :+1:

**Comment by davelopez** [comment]
@ElectronicBlueberry, I think I've addressed all or most of the comments with the following nuances:

> Would it be possible to indicate if a history is deleted/recoverable storage space within the main bar chart view, without hovering, to make them more obvious?

I wanted to make the BarChart somewhat generic so the only information available in a `DataValuePoint` is `id`, `label`, and `value` (size in our case) so unless we pass some more information to `DataValuePoint` I cannot conditionally render this without adding much more complexity. Maybe you have an idea?

---
> It can be quite challenging to hover over these smaller bars. would it be possible to trigger the tooltip on hovering over the column?

I couldn't find an easy way to make the whole column hoverable without too much juggling and even doing so it felt like hovering over an invisible column could make it even more confusing, so I decided to increase the minimum size of the bar "from the bottom" just to make the hovering and click area bigger. Hope this is better.

image

---
> I find this link a bit too hidden. Maybe add a small arrow to the left, or change it to a button

When converting the headings as you suggested, I decided to also align them to the left so it is more consistent with other pages, I think doing so now makes the router link more obvious. What do you think?


Screenshot from 2023-05-17 11-31-22

**Review by ElectronicBlueberry — APPROVED** [review]
We resolved the history hover issue, by making it more obvious that the legend can be clicked. The bar chart overflow was reverted.

**Comment by davelopez** [comment]
Thank you so much @ElectronicBlueberry!

**Pull request merged**